pwndbg/elf.py

```diff
@@ -120,28 +122,55 @@ def reset_ehdr_type_loaded():
     global ehdr_type_loaded
     ehdr_type_loaded = 0
 
+@pwndbg.abi.LinuxOnly()
+def find_elf_magic(pointer, max_pages=1024, search_down=False):
+    """Find the page address which contains the ELF headers
+    by comparing the ELF magic with first 4 bytes.
+
+    Returns:
+        An integer address of ELF page base
+        None if not found within the page limit
+    """
+    base = pwndbg.memory.page_align(pointer)
+    step = pwndbg.memory.PAGE_SIZE
+    if not search_down:
+        step = -step
+
+    max_addr = pwndbg.arch.ptrmask
+
+    for i in range(max_pages):
+        # Make sure address within valid range or gdb will raise Overflow exception
+        if base < 0 or base > max_addr:
+            return None
+
+        try:
+            data = pwndbg.memory.read(base, 4)
+        except gdb.MemoryError:
```

Comment:
Future note: we probably should catch this in `pwndbg.memory.read` and throw a custom exception
(would make it easier to abstract gdb out in the future...)